WIP initial commit for issue 327: Add Swift binding

ScottThomasMiller opened the pull request:
This is my first-ever contribution to an open source project!  Thanks to Andrey for his guidance and for inviting me to contribute these Swift bindings for BrainFlow.  And thanks to William @ OpenBCI for his timely chime-ins on the BrainFlow Slack channel.

BoardShim.swift is fully coded and unit-tested using the synthetic board.  DataFilter.swift is only about 25% complete, but I figured it's OK to include it in this WIP PR.  Both modules are modeled after their Java counterparts.

Other to-do's include:

- bridging header
- libBoardController.dylib and libDataHandler.dylib, built for iOS on arm64, with build instructions
- example Xcode project with unit tests

Andrey1994 commented:
Thanks, @ScottThomasMiller, I will review it soon and will do some minor changes

ScottThomasMiller commented:
Per Andrey, I will move all global functions in BoardShim.swift to static functions within the type definition.

ScottThomasMiller commented:
I am working on the binding for perform_wavelet_transform in DataFilter.swift.  The Java version returns a Pair type.  The internet says a Java Pair is sort of like a Swift dictionary, with key/value pairs.  But in your example code, you seem to use the return value like a normal tuple. Should my binding return a plain vanilla tuple, or a dictionary?

Andrey1994 commented:
Its a tuple of two arrays

ScottThomasMiller commented:
For perform_fft, which returns Complex[], I can think of three options off the top of my head: 

1. Write my own Complex data type in Swift, modeling it after https://introcs.cs.princeton.edu/java/97data/Complex.java.html or similar.
2. Use the Complex data type already defined by Apple in the Swift Numerics package.
3. Simply return a tuple of [Double] arrays.

Option 1 will take some time.  Option 2 will introduce a dependency on a package that is not included in the standard Swift library.  Option 3 forces the caller to do their own numeric processing of complex numbers.

Andrey1994 commented:
How does the process of installing new packages(dependencies) look like in Swift? If it's smth like in python or in java(declare them in files like setup.py or pom.xml and they are downloaded automatically) its ok to introduce Swift Numerics as a dependency.

Andrey1994 commented:
I think we need to add smth like this https://developer.apple.com/documentation/xcode/creating_a_standalone_swift_package_with_xcode and declare it as a dependency in project file

ScottThomasMiller commented:
It's pretty easy.  In my Xcode project for my iOS app, I selected File-> Swift Packages->Add package dependency.  And then I entered the Github URL for the Swift Numerics package: https://github.com/apple/swift-numerics.git.  And then I followed the prompts to assign various dependencies.  Xcode downloaded the package and set the build dependencies per the prompts.  

And then, at the top of DataFilter.swift I added: 

`Import ComplexModule
`

Now I can declare the FFT function as returning an array of Complex Doubles, i.e.:

`static func performFFT (data: inout [Double], startPos: Int32, endPos: Int32, window: Int32) throws -> [Complex<Double>] 
`

I will look into the package bundling, per the link you sent.

ScottThomasMiller commented:
Correction: I have to import the Numerics package itself, not ComplexModule.  Otherwise it seems to be working fine.

Andrey1994 commented:
Its fine. We need to create brainflow package like its described at the link I've shared and add it as a dependency

ScottThomasMiller commented:
DataFilter.swift is 100% coded but only about 10% tested.  I need help with its unit tests, specifically with designing a small handful of hardcoded input data arrays for which we know the expected output.

I am working on the package. It's not easy.  It won't build without the BrainFlow libraries.  I think I need to create a framework for the C++ libs, and then include that framework as a dependency in the package.

Andrey1994 commented:
<< DataFilter.swift is 100% coded but only about 10% tested. I need help with its unit tests, specifically with designing a small handful of hardcoded input data arrays for which we know the expected output.

You should take a look at tests for other languages, you can find them here https://github.com/brainflow-dev/brainflow/tree/master/tests/python for example. Also, need to add it to CI pipelines https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml 

<< I think I need to create a framework for the C++ libs, and then include that framework as a dependency in the package.
No, you need to reuse CMake pipelines as for other languages and copy libs to a correct folder for swift. These libs should be packages together with swift binding

Andrey1994 commented:
About copying libs, there are multiple files like build.cmake in different folders and there are lines like this one https://github.com/brainflow-dev/brainflow/blob/master/src/board_controller/build.cmake#L168 it should be added for swift. Easiest option to find them all is `find . - name build.cmake`

ScottThomasMiller commented:
In band_power_all.py, what are the expected values for band[0] and band[1]?

Andrey1994 commented:
If you use all eeg channels from synthetic board these values should be low

ScottThomasMiller commented:
There's a bug in DataFilter.getAvgBandPowers().  It's returning empty arrays.  I'm looking into it.

ScottThomasMiller commented:
The bug was in my unit test, not in DataFilter.  I updated my unit test, so that now, at the end of it, it does the following:

```
            let bands = try DataFilter.getAvgBandPowers(data: data, channels: EEGchannels,
                                                        samplingRate: samplingRate, applyFilters: true)
            
            let avgSum = bands.0.reduce(0, +)
            let stdSum = bands.1.reduce(0, +)
            XCTAssert((bands.0.count == 5) && (bands.1.count == 5) &&
                      (avgSum > 0) && (avgSum <= 1) && (stdSum > 0) && (stdSum < 10))

```

I ran the test successfully, 10 times in a row.

Andrey1994 commented:
Good, thanks! Let's also add 3rd main component(MLModel) and I will help to update the packaging

ScottThomasMiller commented:
Will do, as soon as I finish testing DataFilter.

Andrey1994 commented:
No, for python it gets them from the shape of ndarray. It's done because of some language-specific implementations of arrays in java and numpy

Andrey1994 commented:
https://github.com/brainflow-dev/brainflow/blob/master/java-package/brainflow/src/main/java/brainflow/DataFilter.java#L440 

here is a comment about it. In c\c++ or in python you can use pointer offsets or slices in java I didn't find anything like that and had to add offsets as parameters

ScottThomasMiller commented:
I can rewrite DataFilter.performFFT() so that it is modeled after the Python, instead of the Java.  Or I can overload it with both signatures.

Andrey1994 commented:
Its up to you

ScottThomasMiller commented:
I completed the initial phase of unit testing for DataFilter, using the Python tests as models.  So far my unit tests cover only the following DataFilter methods:

deTrend
getAvgBandPowers
getBandPower
getCSP
getNearestPowerOfTwo
getPSDwelch
getWindow
performBandpass
performBandstop
performDownsampling
performFFT
performHighpass
performIFFT
performInverseWaveletTransform
performLowpass
performRollingFilter
performWaveletDenoising
performWaveletTransform
removeEnvironmentalNoise

I will start including my unit test modules in the next commit.

Andrey1994 commented:
Is it unit tests or integration tests? We currently use tests as code samples at BrainFlow docs

ScottThomasMiller commented:
These are unit tests for Xcode.  They test only that my bindings return data, and that the output data is different from the input data, except testBandPowerAll() which also compares averages.  I am starting to include negative tests for specific exceptions.

Per your suggestion I followed https://github.com/brainflow-dev/brainflow/tree/master/tests/python.  From that page I implemented the following:

testBandPower
testBandPowerAll
testCSP
testDenoising
testDownsampling
testSignalFiltering
testTransforms
testWindowing

For the CI pipeline should I follow the Examples.html#python page instead?  Is it possible to use one set of tests for both Xcode unit and Github CI?

Andrey1994 commented:
<<For the CI pipeline should I follow the Examples.html#python page instead? Is it possible to use one set of tests for both Xcode unit and Github CI?

Its great that you've added unit tests, I still want to add some unit tests for C++ code and have no time for it.
In CI pipelines we should add both(your unit tests and integration tests like at docs)

ScottThomasMiller commented:
In the CI tests I removed the Python-specific stuff about args parsing and pandas plotting.  Is that stuff absolutely necessary for CI?

Andrey1994 commented:
"In the CI tests I removed the Python-specific stuff about args parsing and pandas plotting. Is that stuff absolutely necessary for CI?"

Yes, because its the only way to test different devices using a single script, also its convenient to have similar tests across languages. And finally, its only in one test, for others we use synthetic board

Andrey1994 commented:
I've checked it, list of remaining tasks:

* add copy for all build files into swift-package/brainflow/lib(I will do it)
* Need to create a project file for swift like here https://developer.apple.com/documentation/xcode/creating_a_standalone_swift_package_with_xcode
* I think compiled libraries should be added into Resource section in this file https://forums.swift.org/t/using-a-dynamically-linked-c-library-dylib-from-spm/47839/2
* Test it on Linux and Windows

ScottThomasMiller commented:
OK.  Sounds good.

So far I've managed to build the dylibs for iOS Simulator using a mix of cmake, sed, and Xcode in interactive mode.  I think there's a command line for the Xcode bit as well.  The toolchain I am using comes from //github.com/leetal/ios-cmake.git. 

I am documenting the process, but it's not yet completely automatable.  My notes are attached.

[StepsForBuildingBrainFlowLibrariesForIosSimulatorOnM1.odt](https://github.com/brainflow-dev/brainflow/files/7232552/StepsForBuildingBrainFlowLibrariesForIosSimulatorOnM1.odt)

Andrey1994 commented:
Also, I've noticed that folder structure doesnt match swift standard. You can run smth like `swift package init --type library`  https://medium.com/developerinsider/create-a-library-using-swift-package-manager-stepwise-tutorial-181604ec8a7

ScottThomasMiller commented:
I've already started playing around with SPM.  I have a package created via "swift package init".  I am trying to make it build the libraries, but so far all I've managed to do is to get it to not throw a bunch of errors.

If you have any luck with it, please let me know.

Andrey1994 commented:
"
I've already started playing around with SPM. I have a package created via "swift package init". I am trying to make it build the libraries, but so far all I've managed to do is to get it to not throw a bunch of errors.

If you have any luck with it, please let me know.
"

You should not build libraries for it you need to just package them.  I will send you a patch which does it in slack

Andrey1994 commented:
Check this link for packaging dylibs https://forums.swift.org/t/using-a-dynamically-linked-c-library-dylib-from-spm/47839/2 and especially 
```
resources: [
        Resource.copy("lib/libBoardController.dylib"),
        ...........
      ],
```

Check setup.py for full list of files needed

ScottThomasMiller commented:
What have I done wrong this time?

Andrey1994 commented:
it seems like smth is broken at codeql action, its not because of your changes and you can ignore it. I will take a look at it

ScottThomasMiller commented:
Yes thanks. I accidentally included them.  They are removed by the latest push.  Also I updated the top-level .gitignore to add *.dylib and swift-package/brainflow/lib/.

The current swift-package is now organized according to the SPM standard, but it depends on prebuilt dylibs, and it violates the rule against mixing source types such as Swift and C++ in the same package.

I think what we are supposed to do is create a separate Framework for the dylibs themselves, one which includes the header files.  We then reference that Framework as a dependency in our Package of Swift bindings.  And then any app which installs the Package gets the Framework automatically. I _think_.

ScottThomasMiller commented:
Will do.  There might also be a command-line equivalent for the Xcode stuff.

Andrey1994 commented:
I am looking at this right now
it seems like we should not copy libs to swift folder, instead we need to create xcframeworks from compiled libraries and use xcframeworks in swift binding
Here is a reference https://github.com/rudedogg/ZigToolsHelp/pull/1/files 
Also, need to rename the package to brainflow from swift-binding

Andrey1994 commented:
xcframework should be creared using installation directory and build.py

ScottThomasMiller commented:
Thanks. I will look into it today.

On Thu, Apr 7, 2022 at 6:49 AM Andrey Parfenov ***@***.***>
wrote:

> xcframework should be creared using installation directory and build.py
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/brainflow-dev/brainflow/pull/342#issuecomment-1091571858>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARPG6FEZ5JRV2T5WPBBUNPLVD24S3ANCNFSM5DXLK6BQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ScottThomasMiller commented:
I am reading: 

https://developer.apple.com/documentation/swift_packages/distributing_binary_frameworks_as_swift_packages

Andrey1994 commented:
yes, its exactly what we need.
CMake will generete libs and headers in installation folder, after that we should run commands like:
```
xcodebuild -create-xcframework \
    -library lib\libBrainFlow.dylib -headers inc \
    -output BrainFlow.xcframework

xcodebuild -create-xcframework \
    -library lib\libBoardController.dylib -headers inc \
    -output BoardController.xcframework
.....
``` 
We can run them from cmake directly after `isntall` or from `build.py`
And use genereted xcframeworks files in swift package

ScottThomasMiller commented:
I created the xcframeworks and the package, and compiled the package in Xcode. So far so good. Xcode created a target for each framework and seems to be happy. I then started a little test app in Xcode and to it added the package. When I try to compile the test app, Xcode complains that it cannot find the headers. I even added a bridging header. I continue to research.

ScottThomasMiller commented:
Here's how I created the xcframeworks:

0) clone the repo
1) cd brainflow/tools
2) python3 build.py
3) cd ../installed
4) 
xcodebuild -create-xcframework -library lib/libBoardController.dylib -headers inc -output BoardController.xcframework
xcodebuild -create-xcframework -library lib/libBrainBitLib.dylib  -output BrainBitLib.xcframework
xcodebuild -create-xcframework -library lib/libDataHandler.dylib  -output DataHandler.xcframework
xcodebuild -create-xcframework -library lib/libGanglionLib.dylib  -output GanglionLib.xcframework
xcodebuild -create-xcframework -library lib/libMLModule.dylib  -output MLModule.xcframework
xcodebuild -create-xcframework -library lib/libMuseLib.dylib  -output MuseLib.xcframework
xcodebuild -create-xcframework -library lib/libneurosdk-shared.dylib  -output neurosdk-shared.xcframework

I specified the headers only in the first framework because otherwise Xcode complains about duplicate headers when I try to compile the package. I figured that would be easier than separating each header into its respective framework due to interdependencies amongst the headers.

ScottThomasMiller commented:
The package looks like this:

// swift-tools-version:5.3
import PackageDescription

let package = Package(
    name: "BrainFlow",
    platforms: [
        .macOS(.v10_14)
    ],
    products: [
        // Products define the executables and libraries a package produces, and make them visible to other packages.
        .library(
            name: "BrainFlow",
            targets: ["BrainFlow", "BoardController", "BrainBitLib", "DataHandler",
                      "GanglionLib", "MLModule", "MuseLib", "neurosdk-shared" ])
    ],
    dependencies: [
        // Dependencies declare other packages that this package depends on.
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages this package depends on.
        .target(
            name: "BrainFlow"
        ),
        .binaryTarget(
            name: "BoardController",
            path: "BoardController.xcframework"
        ),
        .binaryTarget(
            name: "BrainBitLib",
            path: "BrainBitLib.xcframework"
        ),
        .binaryTarget(
            name: "DataHandler",
            path: "DataHandler.xcframework"
        ),
        .binaryTarget(
            name: "GanglionLib",
            path: "GanglionLib.xcframework"
        ),
        .binaryTarget(
            name: "MLModule",
            path: "MLModule.xcframework"
        ),
        .binaryTarget(
            name: "MuseLib",
            path: "MuseLib.xcframework"
        ),
        .binaryTarget(
            name: "neurosdk-shared",
            path: "neurosdk-shared.xcframework"
        ),
        .testTarget(
            name: "BrainFlowTests",
            dependencies: ["BrainFlow"]
        )
    ]
)

Andrey1994 commented:
thats correct, but I think 4 should be done automatically from the script

ScottThomasMiller commented:
Progress! Check it out:

<img width="1078" alt="image" src="https://user-images.githubusercontent.com/73297684/162632500-d3c47c0f-62b4-471b-ad4c-593490d4654b.png">

ScottThomasMiller commented:
I had to manually configure the bridging header (bridge.h in the example above) separate from the package. I will try to include it within the package, and then see if Xcode can find it there instead. 

Also I used arm64 binaries, so a next step is to rebuild the libs as universal binaries. 

And I had to remove the neurosdk-shared target from the package (_not_ shown above), because the dynamic linker complained that it couldn't find libneurosdk.dylib. Theoretically I can replace it with a URL that points to BrainBit's downloadable framework. Please keep in mind that if I include both libneurosdk-shared.dylib and libneurosdk.dylib (which I currently must do for my BCI app), then dyld prints warnings about duplicate symbols.

Andrey1994 commented:
<<Also I used arm64 binaries, so a next step is to rebuild the libs as universal binaries.

CI builds it automatically as universal libs, in order to merge this PR its not important

ScottThomasMiller commented:
I moved bridge.h into the package, under the Sources folder, and updated the Xcode config accordingly. It works fine.

<img width="1465" alt="image" src="https://user-images.githubusercontent.com/73297684/162840464-65b01693-a990-4138-92b2-519c0abc0c1c.png">

Andrey1994 commented:
https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml#L62 check this, it uses cmake directly instead build.py but I think using build.py the same can be done.

Need to use ninja and specify deployment target

ScottThomasMiller commented:
That worked. Thanks. I did:

```
export BRAINFLOW_VERSION=4.9.0
export GITHUB_WORKSPACE=/Users/scottmiller/git/brainflow
cmake -G Ninja -DBRAINFLOW_VERSION=$BRAINFLOW_VERSION -DBUILD_BLUETOOTH=ON -DBUILD_TESTS=ON -DWARNINGS_AS_ERRORS=ON \
  -DCMAKE_OSX_ARCHITECTURES="arm64;x86_64" -DCMAKE_OSX_DEPLOYMENT_TARGET=10.13 -DCMAKE_INSTALL_PREFIX=$GITHUB_WORKSPACE/installed \
  -DCMAKE_BUILD_TYPE=Release ..
ninja
ninja install

```

ScottThomasMiller commented:
I cannot replace libneurosdk.dylib with a remote URL reference, because the package on BrainBit website is a framework, not an xcframework. I could try converting it into an xcframework, but I am not sure that is how we want to go.

So I added libneurosdk-shared.dylib back into the package. But now when I try to run my test app I get the following error:

```
dyld[18110]: Library not loaded: @rpath/libneurosdk.dylib
  Referenced from: /Users/scottmiller/Library/Developer/Xcode/DerivedData/testlibs-cfuquabhzufcshfcwkntzgsmlbyl/Build/Products/Debug/testlibs.app/Contents/MacOS/testlibs
  Reason: tried: '/Users/scottmiller/Library/Developer/Xcode/DerivedData/testlibs-cfuquabhzufcshfcwkntzgsmlbyl/Build/Products/Debug/libneurosdk.dylib' (no such file), '/usr/lib/system/introspection/libneurosdk.dylib' (no such file), '/Users/scottmiller/Library/Developer/Xcode/DerivedData/testlibs-cfuquabhzufcshfcwkntzgsmlbyl/Build/Products/Debug/testlibs.app/Contents/MacOS/../Frameworks/libneurosdk.dylib' (no such file), '/Users/scottmiller/Library/Developer/Xcode/DerivedData/testlibs-cfuquabhzufcshfcwkntzgsmlbyl/Build/Products/Debug/testlibs.app/Contents/MacOS/../Frameworks/libneurosdk.dylib' (no such file), '/usr/local/lib/libneurosdk.dylib' (no such file), '/usr/lib/libneurosdk.dylib' (no such file)
(lldb) 

```

Andrey1994 commented:
At this moment I dont really care about 3rd party libs, I want to see core boards integrated(boards that do not require extra libs from other companies) first and it will be good to merge this PR

Andrey1994 commented:
technically we can even drop libneurosdk if its a problem and write own BLE implementation(like in brainbit bled)

ScottThomasMiller commented:
I removed libneurosdk-shared and rebuilt the package. In my package-test app I tested the synth board successfully, but the CYTON_BOARD returns the following error:

```
[2022-04-20 06:36:42.441] [board_logger] [info] prepare session
[2022-04-20 06:36:42.441] [board_logger] [info] incoming json: {
  "serial_port" : "/dev/cu.usbserial-DM0258EJ",
  "mac_address" : "",
  "serial_number" : "",
  "ip_protocol" : 0,
  "ip_port" : 0,
  "timeout" : 0,
  "file" : "",
  "other_info" : "",
  "ip_address" : ""
}
[2022-04-20 06:36:42.441] [board_logger] [trace] Board object created 0
[2022-04-20 06:36:42.441] [board_logger] [info] opening port /dev/cu.usbserial-DM0258EJ
[2022-04-20 06:36:42.442] [board_logger] [error] Make sure you provided correct port name and have permissions to open it(run with sudo/admin). Also, close all other apps using this port.
[2022-04-20 06:36:42.442] [board_logger] [error] failed to prepare session
[2022-04-20 06:36:42.442] [board_logger] [error] Error code: UNABLE_TO_OPEN_PORT_ERROR

```

No other apps are using the port. I do not get the error from my BCI app, in which I embed the dylibs directly as individual frameworks.

ScottThomasMiller commented:
I tried archiving the app and running it as root. I also tried granting Developer Tools permissions to the archived app via Security & Privacy Settings. No luck. So I posted the issue to the Swift forum:

https://forums.swift.org/t/cannot-access-usb-serial-port-via-binary-target-within-a-package/56851

Andrey1994 commented:
wow, I have no idea what can be the issue here

ScottThomasMiller commented:
I also posted the same question to the Apple Developer forum. If necessary I will post it to stackoverflow as well, but first I will strip it down to the core libraries, and include samples from both my code and the relevant C++ code. If all else fails, then I will open a support ticket with Apple.

Which do you consider the core libraries? And which line of C++ code opens the serial port for the CYTON board?

Andrey1994 commented:
BoardController DataHandler MLModule as soon as these 3 work we can merge this PR. Also, bindings call methods **only** from these 3 dylibs, other dylibs are loaded by BoardController if needed and bindings do not communicate with them directly

src/utils/os_serial is the file with serial port read write methods, they are called from src/board_controoler/openbci/openbci_serial_board.cpp

ScottThomasMiller commented:
The Mighty Quinn replied to my post on the Swift forum and asked for additional details, which I promptly provided.

ScottThomasMiller commented:
Fixed. It wasn't a Swift Package issue. I had to disable the app sandbox via the Entitlements File. Disabling the app sandbox is not a great idea because Apple will not let you distribute your app via their App Store without sandboxing (according to the internet.)

The Package is actually working fine.

For now I will mention the sandbox issue when I write the readme for the Swift bindings. When I have time I will try to figure out a better fix.

ScottThomasMiller commented:
My swift-bindings branch I think is too far behind the main branch. For example my branch has set_log_level() instead of the newer set_log_level_board_controller(). I am trying to update my branch with the latest from the main branch but I cannot figure out the git syntax. How should I proceed?

Andrey1994 commented:
run this:
```
git remote add upstream https://github.com/brainflow-dev/brainflow
git fetch upstream
git merge upstream/master
at this moment git will complain about confilcts, you have to fix it manually in each file and run git add for each of these files
git merge --continue
```

ScottThomasMiller commented:
The conflicts are in several .cmake files, and the .gitignore file. I have no idea how to fix them. Would it be easier to start from scratch with a new branch?

Andrey1994 commented:
They should be easy to solve but if its easier for you to start a new one its also ok

ScottThomasMiller commented:
To fix the conflicts should I simply overwrite the offending .cmake files with the latest versions from master?

Andrey1994 commented:
https://github.com/brainflow-dev/brainflow/pull/342#issuecomment-1120208709

I suppose in your branch the only change in cmake files is copypaste of dylibs to swift folder and since we decided to use xcframework its not needed, so I think you can drop your changes in cmake files and just use the latest from master.

ScottThomasMiller commented:
The Package works but could be more convenient. It's not completely self-contained, and I am still working on improving that aspect. For now the bindings need to be copied from the Package into the target app's group in Xcode, and the bridging header still needs to be configured. Ideally it would be installable via Github URL, which currently is the ultimate in convenience.

Andrey1994 commented:
Great news and thanks a lot! I will review everything tomorrow
About failed CI pipeline you probably need to install clang-format and https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#code-style resave the file. Maybe it complains about trailing spaces or smth

ScottThomasMiller commented:
So if I understand the CI error, my local copy of src/board_controller/inc/board_info_getter.h	does not match its clang-format output?

ScottThomasMiller commented:
There does appear to be a trailing space at the end of the line in my local copy.

Andrey1994 commented:
thanks and CI is back to green!

Could you also update docs with instructions about swift package? https://brainflow.readthedocs.io/en/stable/BrainFlowDev.html#instructions-to-build-docs-locally 

Need to add swift package installation instructions to https://github.com/brainflow-dev/brainflow/blob/master/docs/BuildBrainFlow.rst 

A few sentences about user api, no need to autogenerate it for now https://github.com/brainflow-dev/brainflow/blob/master/docs/UserAPI.rst 

And links to examples https://github.com/brainflow-dev/brainflow/blob/master/docs/Examples.rst

Andrey1994 reviewed (changes requested):
looks good, I have some minor comments but first of all in addition to docs we need to add tests to CI pipelines. Check https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml Let me know if you want me to help with smth

ScottThomasMiller commented:
I am working on the docs now. The CI tests are in swift-package/BrainFlow/Tests/BrainFlowTests/BrainFlowCItests.swift. I don't have any CI experience so I could use some help with that.

Andrey1994 commented:
no problem, I can help with CI, roughly speaking you need to add commands to the yaml file, if exit code of the command is not zero pipeline will be failed

ScottThomasMiller commented:
I hit a snag testing my instructions. It's fixed now. I had to rearrange the package a little. 

I am still working on splitting out the examples into individual files.

Andrey1994 commented:
good, thanks for update

Andrey1994 commented:
I think in the last commit you removed some files but didnt add renamed files

ScottThomasMiller commented:
I moved the bindings to swift-package/BrainFlow/Sources/BrainFlowBindings. Is that what you are referring to?

Andrey1994 commented:
> I moved the bindings to swift-package/BrainFlow/Sources/BrainFlowBindings. Is that what you are referring to?

yes, no worries, I somehow didnt see some files

ScottThomasMiller commented:
For the examples I got lazy. Instead of splitting them up into separate functions, I kept them all in the BrainFlowCITests.swift file. They are also the CI tests. If you insist, I will break them out into separate files. 

Regarding the CI tests: which build tools are available to us? Won't we need something like xcrun or xcodebuild?

Andrey1994 commented:
<<Regarding the CI tests: which build tools are available to us? Won't we need something like xcrun or xcodebuild?

Everything that is preinstalled in VM + its possible to install smth else via predefined actions or from homebrew directly

Here is more info https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md

ScottThomasMiller commented:
OK I have something working for CI that is ready to push. I added a new Xcode project in swift-package/BrainFlowCI. The command lines to run the unit tests look like this:

cd swift-package/BrainFlowCI
xcodebuild \
  -project BrainFlowCI.xcodeproj \
  -scheme BrainFlowCI \
  -sdk macosx12.3 \
  -destination 'platform=OS X' \
  test

Andrey1994 commented:
> I have something working for CI that is ready to push. I added a new Xcode project in swift-package/BrainFlowCI. The command lines to run the unit tests look like this:

nice, feel free to push. 

Also, there were some breaking changes in brainflow 5.0.0, in MLModule predict now should return an array, in bandpass\bandstop now we use start and stop freqs and wavelets are enums now

ScottThomasMiller commented:
> Also, there were some breaking changes in brainflow 5.0.0, in MLModule predict now should return an array, in bandpass\bandstop now we use start and stop freqs and wavelets are enums now

CI tests pushed, and merge conflicts resolved. I am looking at the 5.0.0 changes now.

Andrey1994 commented:
https://github.com/brainflow-dev/brainflow/pull/441 I tried to write them in description for this pr

ScottThomasMiller commented:
How do I merge the 5.0.0 files with those in my branch? Like this? 

https://stackoverflow.com/questions/33484153/branch-is-behind-main

Andrey1994 commented:
> 

in a folder where you cloned it run:

```
git remote add upstream https://github.com/brainflow-dev/brainflow
git fetch upstream
git merge upstream/master
```

It will show you some conflicts, need to solve them and run:
```
git add %file with conflict%
git merge --continue
```

ScottThomasMiller commented:
I'm working my way down the list. I am at this one:

> java: add method overloading for enums in addition to ints. e.g: BoardIds board_id = BoardIds.SYNTHETIC_BOARD, before it had to be int board_id = BoardIds.SYNTHETIC_BOARD.get_code()

I think we get that one for free with Swift. I can already declare smth like:

`var boardId: BoardIds = .SYNTHETIC_BOARD`

Do I need to do anything for this one?

Andrey1994 commented:
No, it was only for java and its speciffic to enums in java, in other languages they are integers from the box

ScottThomasMiller commented:
I see what I need to do for the last one, but do I need to worry about the next-to-last one:

> cpp: methods for fft now return size

Andrey1994 commented:
Its also not relevant

Andrey1994 commented:
good, I will review it one more time and this time more carefully. I want to merge your changes before #510 because it will require changes in Swift binding also and if we merge this PR first I will port them to swift by myself.

About metrics and classifiers, indeed not all combinations are supported, could you tell me which one you tried? Exception should be this one https://github.com/brainflow-dev/brainflow/blob/master/src/ml/ml_module.cpp#L67 and probably I need to add log message here. If you see an exception in json its a little strange

ScottThomasMiller commented:
This one works:

`let concentrationParams = BrainFlowModelParams(metric: .MINDFULNESS, classifier: .DYN_LIB_CLASSIFIER)
`

This one does not work:

`let relaxationParams = BrainFlowModelParams(metric: .RESTFULNESS, classifier: .ONNX_CLASSIFIER)
`

ScottThomasMiller commented:
Correction: none of them are working. I was mislead by Xcode. Not even the ones that should work, for example mindfulness+default.

ScottThomasMiller commented:
It was my bug. Fixed. Upgrades complete, including the CI tests.

Andrey1994 commented on swift-package/BrainFlow/Sources/BrainFlowBindings/DataFilter.swift:
```diff
@@ -0,0 +1,561 @@
+        do {
+            try setLogLevel (.LEVEL_OFF) }
+        catch {
+            throw error
```
here the same

Andrey1994 commented on swift-package/BrainFlow/Sources/BrainFlowBindings/DataFilter.swift:
```diff
@@ -0,0 +1,561 @@
+    /**
+     * perform wavelet transform
+     *
+     * @param wavelet supported vals:
```
this comment is not relevant anymore, now its enum, not a string

Andrey1994 commented on swift-package/BrainFlow/Sources/BrainFlowBindings/MLModule.swift:
```diff
@@ -0,0 +1,101 @@
+    
+    let modelParams: BrainFlowModelParams
+    
+    static func setLogLevel(_ logLevel: LogLevels) throws {
```
`log_message` method is missed

Andrey1994 commented on swift-package/BrainFlowCI/BrainFlowCI.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,540 @@
+// !$*UTF8*$!
```
do we need this file?

Andrey1994 commented on swift-package/BrainFlowCI/BrainFlowCI.xcodeproj/project.xcworkspace/contents.xcworkspacedata:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
and this one

Andrey1994 commented on swift-package/BrainFlowCI/BrainFlowCI.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
and this one and a few more files below

Andrey1994 commented on swift-package/build_package_macOS.sh:
```diff
@@ -0,0 +1,20 @@
+
```
I think its better to move it to build.py

ScottThomasMiller replied:
ok

ScottThomasMiller replied:
ok

Andrey1994 commented on swift-package/BrainFlowCI/BrainFlowCI/main.swift:
```diff
@@ -0,0 +1,11 @@
+//
```
do we need this?

ScottThomasMiller replied:
not 100% sure

ScottThomasMiller replied:
Maybe.

Andrey1994 commented on swift-package/BrainFlowCI/RunCI.sh:
```diff
@@ -0,0 +1,7 @@
+
```
I would add it to CI as commands in yaml, dont see the need for shell script

ScottThomasMiller replied:
ok

ScottThomasMiller replied:
Hi. I took a break to work on trying to build the BrainFlow dylibs using Xcode so that they are acceptable to the App Store. I have made progress but I'm not quite there yet. I will return to focusing on the CI cmds in a few days. Later I think I should open a new issue for the App Store build.

Is it easy to build static .a libraries for BrainFlow, instead of .dylibs?

Andrey1994 reviewed (changes requested):
Great job!

I wrote some comments but it looks really good.

One more question, files like DS_Store, what is it and do we really need them? I think they are pushed by mistake, at least they are in gitignore..

Svm model should also be removed.

About examples: I think we should split unittests and examples and make examples consistent with other languages

Amd the most important one: I dont see changes in CI jobs. You need to build swift package and run some tests here https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml here is an example how to build smth only for macos https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml#L54 here is an example how to run tests https://github.com/brainflow-dev/brainflow/blob/master/.github/workflows/run_unix.yml#L54

Andrey1994 commented:
also, probably I missed it but I didnt see methods like `release_all_sessions` they should be in all 3 classes

ScottThomasMiller commented:
I seem to have a lot of accidentally pushed flotsam in my branch. I apologize and will clean it all up shortly.

ScottThomasMiller commented on docs/Examples.rst:
```diff
@@ -492,3 +492,9 @@ Notebooks
+Swift
+------
+
+.. literalinclude:: ../swift-package/BrainFlow/Tests/BrainFlowTests/BrainFlowCITests.swift
```
ok

ScottThomasMiller replied:
ok

ScottThomasMiller commented:
Everything under swift-package/BrainFlowCI)/BrainFlowCI.xcodeproj/ was created by Xcode and I assume should be in the repo.

ScottThomasMiller commented on swift-package/BrainFlow/Package.resolved:
```diff
@@ -0,0 +1,16 @@
+{
```
no

ScottThomasMiller commented on swift-package/BrainFlow/BrainFlow.h:
```diff
@@ -0,0 +1,13 @@
+//  Created by Scott Miller on 4/9/22.
+//
+
+#ifndef BrainFlow_h
```
ok

ScottThomasMiller commented on docs/BuildBrainFlow.rst:
```diff
@@ -127,6 +127,47 @@ Rust
+- Copy the BrainFlow folder and its entire contents from /Users/myusername/brainflow/swift-package to /Users/myusername/myproject.
+- In Xcode select the top-level project and then click File->Add Packages...Add Local...
+- Select /Users/myusername/myproject/BrainFlow. Click Open.
+- Drag-and-drop the BrainFlowBindings subfolder from myproject/Packages/BrainFLow/Sources into myproject/myapp. 
```
ok

Andrey1994 replied:
maybe instead github artifacts we can publish to aws(run_unix workflow does it currently) I can take care of publishing and packaging but need to build package inside CI first

ScottThomasMiller commented on docs/BuildBrainFlow.rst:
```diff
@@ -127,6 +127,47 @@ Rust
+- in a shell such as iTerm2:
+
+  cd /Users/myusername/brainflow/swift-package
+  ./build_package_macOS.sh
```
I will take a look.

ScottThomasMiller replied:
For the examples, do you want something like this:

```
>ls -l ./r_package/examples
total 80
-rw-r--r--  1 scottmiller  staff   957 Jun 24 06:17 band_power.R
-rw-r--r--  1 scottmiller  staff   706 Jun 24 06:17 band_power_all.R
-rw-r--r--  1 scottmiller  staff   356 Jun 24 06:17 brainflow_get_data.R
-rw-r--r--  1 scottmiller  staff   555 Jun 24 06:17 denoising.R
-rw-r--r--  1 scottmiller  staff   586 Jun 24 06:17 downsampling.R
-rw-r--r--  1 scottmiller  staff  1090 Jun 24 06:17 eeg_metrics.R
-rw-r--r--  1 scottmiller  staff   384 Jun 24 06:17 markers.R
-rw-r--r--  1 scottmiller  staff   541 Jun 24 06:17 serialization.R
-rw-r--r--  1 scottmiller  staff   708 Jun 24 06:17 signal_filtering.R
-rw-r--r--  1 scottmiller  staff   709 Jun 24 06:17 transforms.R

```

Andrey1994 replied:
yes, exactly. R is not the best example but you got the idea. Would be better to use python as a reference. You can also find them here https://github.com/brainflow-dev/brainflow/blob/master/docs/Examples.rst(will also need to add swift examples to this page)

Andrey1994 commented:
> Everything under swift-package/BrainFlowCI)/BrainFlowCI.xcodeproj/ was created by Xcode and I assume should be in the repo.



> Everything under swift-package/BrainFlowCI)/BrainFlowCI.xcodeproj/ was created by Xcode and I assume should be in the repo.

I thought some of shell scripts\commands generate it or xcode creates it automatically, in this case probably we dont need it in the repo

Andrey1994 commented:
I had to merge one more change which will require swift package update. Here you can find more info https://brainflow.org/2022-07-15-brainflow-5-1-0/  and https://github.com/brainflow-dev/brainflow/pull/510 

CI is especially important now because I relocated and have no access to macos anymore... so its the only way for me to see that it actually works and passes tests
